File: cses/07_mathematics/2207_grundys-game/solution.py
import sys


def main():
    # P-position cuối cùng của trò chơi là 1222; chọn ngưỡng B lớn hơn một chút.
    B = 1300
    # Tính giá trị Grundy g(k) theo công thức truy hồi Sprague-Grundy.
    # g[1] = g[2] = 0 vì không thể tách thành hai đống khác kích thước.
    g = [0] * (B + 1)
    for k in range(3, B + 1):
        m = (k - 1) // 2  # a chạy 1..m để bảo đảm a < k-a (hai đống khác nhau)
        # Tập các nimber đạt được khi tách đống k thành a và k-a.
        seen = set()
        add = seen.add
        for a in range(1, m + 1):
            add(g[a] ^ g[k - a])
        # g(k) = mex của seen (số nguyên không âm nhỏ nhất chưa xuất hiện).
        mex = 0
        while mex in seen:
            mex += 1
        g[k] = mex

    # Đọc toàn bộ input một lần cho nhanh.
    data = sys.stdin.buffer.read().split()
    t = int(data[0])
    out = []
    for i in range(1, t + 1):
        n = int(data[i])
        if n <= B:
            # g(n)==0 là P-position: người đi thua -> "second"; ngược lại "first".
            out.append("second" if g[n] == 0 else "first")
        else:
            out.append("first")  # vượt 1222 thì không còn P-position nào
    sys.stdout.write("\n".join(out))
    if out:
        sys.stdout.write("\n")


main()

File: cses/07_mathematics/2207_grundys-game/solution.cpp
#include <iostream>
#include <cstddef>
#include <unordered_set>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    // P-position cuối cùng của trò chơi này là 1222; tính g(k) vượt qua ngưỡng
    // đó một chút là đủ để tái tạo toàn bộ P-position mà không cần hard-code.
    constexpr std::size_t LIMIT = 1300;
    std::vector<int> grundy(LIMIT + 1, 0);
    // Tính giá trị Grundy theo công thức truy hồi Sprague-Grundy.
    for (std::size_t coins = 3; coins <= LIMIT; ++coins) {
        // Tập các nimber đạt được khi tách đống coins thành left và coins-left
        // (left < coins-left để bảo đảm hai đống khác kích thước).
        std::unordered_set<int> reachable;
        for (std::size_t left = 1; left < coins - left; ++left) {
            reachable.insert(grundy[left] ^ grundy[coins - left]);
        }

        // g(coins) = mex của tập reachable (số nguyên không âm nhỏ nhất vắng mặt).
        int mex = 0;
        while (reachable.count(mex) != 0U) {
            ++mex;
        }
        grundy[coins] = mex;
    }

    int tests;
    std::cin >> tests;
    while (tests-- > 0) {
        std::size_t coins;
        std::cin >> coins;
        // g(n)==0 (chỉ có thể xảy ra khi n <= LIMIT) là P-position: người đi
        // sẽ thua nên đáp án là "second"; ngược lại "first".
        const bool losing = coins <= LIMIT && grundy[coins] == 0;
        std::cout << (losing ? "second" : "first") << '\n';
    }
    return 0;
}
